crates/core/action-context/src/lib.rs

```diff
@@ -26,13 +26,59 @@ impl TargetState {
     }
 }
 
+#[derive(Clone, Debug, Default)]
+pub struct TaskNamedMutexes {
+    mutexes: Arc<std::sync::Mutex<FxHashMap<String, Arc<tokio::sync::Mutex<()>>>>>,
+}
+
+impl<'de> Deserialize<'de> for TaskNamedMutexes {
+    fn deserialize<D>(_: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        // We don't care about unserializing this, but we need to satisfy the
+        // trait requirements.
+        Ok(TaskNamedMutexes::new())
+    }
+}
+
+impl Serialize for TaskNamedMutexes {
+    fn serialize<S>(&self, _serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        // We don't care about serializing this, but we need to satisfy the
+        // trait requirements.
+        FxHashMap::<String, ()>::default().serialize(_serializer)
+    }
+}
+
+impl TaskNamedMutexes {
+    fn new() -> Self {
+        TaskNamedMutexes {
+            mutexes: Arc::new(std::sync::Mutex::new(FxHashMap::default())),
+        }
+    }
+
+    pub fn get(&self, name: &str) -> Arc<tokio::sync::Mutex<()>> {
+        // TODO: Check how to remove that `unwrap`
+        let mut mutexes = self.mutexes.lock().unwrap();
+        if !mutexes.contains_key(name) {
+            mutexes.insert(name.to_string(), Arc::new(tokio::sync::Mutex::new(())));
+        }
+        mutexes.get(name).unwrap().clone()
+    }
+}
+
 #[derive(Clone, Debug, Default, Deserialize, Serialize)]
 #[serde(rename_all = "camelCase")]
 pub struct ActionContext {
     pub affected_only: bool,
 
     pub initial_targets: FxHashSet<TargetLocator>,
 
+    pub named_mutexes: TaskNamedMutexes,
```

Comment:
Probably wrap with `#[serde(skip)]` instead of the impls above?

It feels weird putting this on the context, but there's no other good place at the moment.

Reply:
Ah, I didn't know about `#[serde(skip)]`, that makes much more sense than the hack I implemented.

It's been like 2 years since my last meaningful coding sessions in Rust 😅 .